sanitycheck: Add device tree query function dt_chosen_enabled()

sidcha opened the pull request:
To filter by device tree chosen property, add device tree query function
dt_chosen_enabled(name) that returns true if there is a device tree
chosen property for the given name.

Signed-off-by: Siddharth Chandrasekaran <siddharth@embedjournal.com>

erwango commented on samples/subsys/mgmt/osdp/sample.yaml:
```diff
@@ -4,7 +4,7 @@ sample:
   sample.mgmt.osdp:
     tags: osdp
-    filter: dt_compat_enabled_with_alias("gpio-leds", "led0") and CONFIG_SERIAL
+    filter: dt_compat_enabled_with_alias("gpio-leds", "led0") and dt_alias_exists("rng") and CONFIG_SERIAL
```
Ok, this might work to exclude boards, but I don't see any board having any `rng` alias today:
`$ grep aliases -A 10 boards/*/*/*.dts | grep rng` doesn't return anything

cbsiddharth replied:
Sorry, my bad. At the boards level, I think we must filter by dt_chosen "zephyr,entropy".

cbsiddharth replied:
Can you check now?

pfalcon commented:
@cbsiddharth, I'd say, you should address this on source code level [too]. Specifically, `config OSDP_SC_ENABLED` in Kconfig should `depend` on something similar to what `z_impl_sys_csrand_get` itself depends on.

sidcha committed:
sanitycheck: Add device tree query function dt_chosen_enabled()

To filter by device tree chosen property, add device tree query function
dt_chosen_enabled(name) that returns true if there is a device tree
chosen property for the given name.

Signed-off-by: Siddharth Chandrasekaran <siddharth@embedjournal.com>

galak reviewed (changes requested):
This should be dt_chosen_enabled to match `scripts/kconfig/kconfigfunctions.py`

galak commented:
Agreed with @pfalcon that the fix here is on the subsys Kconfig side

cbsiddharth commented:
> This should be dt_chosen_enabled to match `scripts/kconfig/kconfigfunctions.py`

Done.

cbsiddharth commented:
> @cbsiddharth, I'd say, you should address this on source code level [too]. Specifically, `config OSDP_SC_ENABLED` in Kconfig should `depend` on something similar to what `z_impl_sys_csrand_get` itself depends on.

@pfalcon, @galak, I get into a dependency chain when I try to depend on `ENTROPY_HAS_DRIVER`  as it is done in other places because`OSDP` selects `ENTROPY_GENERATOR`. Can you point me to another subsystem that does this right?

pfalcon commented:
> @pfalcon, @galak, I get into a dependency chain when I try to depend on ENTROPY_HAS_DRIVER (as it is done in other places) becauseOSDP selects ENTROPY_GENERATOR. Can you point me to another subsystem that does this right?

Let's first make explicit what's the desired effect. It is so that if a user does `samples/subsys/mgmt/osdp; west build --board=mps2_an385` (also, a bunch of other boards), they don't get an error like `undefined reference to `z_impl_sys_csrand_get'`. Which, I hope you agree, is a confusing outcome.

That's why the comment that this patch is not enough - it will preclude `sanitycheck` tool from selecting this sample, but not a user to do that manually.

So, the desired effect for a(ny) sample is that: it worked for as long as it can, and if it truly can't, to fail with clear error message, explaining a user what's wrong/how to fix it.

Back to your question above: well, that's why, to the best of my knowledge, using too much of `select` is not recommended. It's recommended to `depends on` instead.

Sorry for not providing a more specific suggestion to fix it - I don't have a specific template it mind. Again, the point is to preclude samples (and more generally, user applications) from failing with confusing link errors, so any suitable means towards that end should be good.

ceolin commented:
> > > @cbsiddharth, I'd say, you should address this on source code level [too]. Specifically, `config OSDP_SC_ENABLED` in Kconfig should `depend` on something similar to what `z_impl_sys_csrand_get` itself depends on.
> > 
> > 
> > @pfalcon, @galak, I get into a dependency chain when I try to depend on `ENTROPY_HAS_DRIVER` (as it is done in other places) because`OSDP` selects `ENTROPY_GENERATOR`. Can you point me to another subsystem that does this right?
> 
> Aren't we looking it from the wrong perspective ? Instead of filtering the sample, shouldn't we enable `OSDP_SC_ENABLED` only if there is a truly random number generator available ? Looking osdp Kconfig, it seems that is also wrongly depending in the entropy generator while it should be depending on the random subsys ` CSPRING_ENABLED`

@cbsiddharth  could you take a look in this approach:
https://github.com/zephyrproject-rtos/zephyr/pull/29137

cbsiddharth commented:
> @cbsiddharth could you take a look in this approach:
> #29137

@ceolin, Thank you.

nashif commented:
looks like we have this in #30166
